From stable-bounces@linux.kernel.org  Wed Mar 12 15:05:15 2008
Message-ID: <47CF31CE.5020106@redhat.com>
Date: Wed, 05 Mar 2008 18:50:38 -0500
From: Chuck Ebbert <cebbert@redhat.com>
To: linux-stable <stable@kernel.org>
Cc: "H. Peter Anvin" <hpa@zytor.com>
Subject: x86: don't use P6_NOPs if compiling with CONFIG_X86_GENERIC

From: H. Peter Anvin <hpa@zytor.com>
Commit: 959b3be64cab9160cd74532a49b89cdd918d38e9

x86: don't use P6_NOPs if compiling with CONFIG_X86_GENERIC

P6_NOPs are definitely not supported on some VIA CPUs, and possibly
(unverified) on AMD K7s.  It is also the only thing that prevents a
686 kernel from running on Transmeta TM3x00/5x00 (Crusoe) series.

The performance benefit over generic NOPs is very small, so when
building for generic consumption, avoid using them.

Signed-off-by: H. Peter Anvin <hpa@zytor.com>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
[cebbert@redhat.com: backport to 2.6.24]
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

 include/asm-x86/processor_32.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/include/asm-x86/processor_32.h
+++ b/include/asm-x86/processor_32.h
@@ -714,7 +714,8 @@ static inline unsigned int cpuid_edx(uns
 #define ASM_NOP8 K7_NOP8
 #elif defined(CONFIG_M686) || defined(CONFIG_MPENTIUMII) || \
       defined(CONFIG_MPENTIUMIII) || defined(CONFIG_MPENTIUMM) || \
-      defined(CONFIG_MCORE2) || defined(CONFIG_PENTIUM4)
+      defined(CONFIG_MCORE2) || defined(CONFIG_PENTIUM4) && \
+      !defined(CONFIG_X86_GENERIC)
 #define ASM_NOP1 P6_NOP1
 #define ASM_NOP2 P6_NOP2
 #define ASM_NOP3 P6_NOP3
